src/function_classes/retrieval.py

```diff
@@ -1,22 +1,40 @@
 import torch
 import torch.distributions as D
 
+from torch import Tensor
+
 from core import FunctionClass
+from utils import CombinedDistribution
 
 class Retrieval(FunctionClass):
 
     def _init_param_dist(self) -> D.Distribution:
-        return D.Categorical(torch.Tensor([1]))
-    
-    def evaluate(self, x_batch: torch.Tensor, *params: torch.Tensor) -> torch.Tensor:
-        assert x_batch.shape[0] == self.batch_size and x_batch.shape[2] == self.x_dim
-        assert x_batch.shape[1] % 2 == 1
-
-        keys_b, values_b = x_batch[:, :-1:2, :], x_batch[:, 1::2, :]
-        query_b = x_batch[:, -1, :].unsqueeze(dim=1)
-        inner_products = torch.bmm(query_b, keys_b.transpose(1, 2)).squeeze(1)
-        _, retrieval_inds = torch.max(inner_products, dim=1)
-        retrieval_inds = retrieval_inds.view(-1, 1, 1).expand(-1, -1, values_b.size(-1))
-        y_batch = torch.gather(values_b, 1, retrieval_inds).squeeze(1)
-        
-        return y_batch
+        return CombinedDistribution(
+            self.x_dist,
+            D.Categorical(
+                torch.ones([self.sequence_length]) 
+                / self.sequence_length
+            )
+        )
```

Comment:
if you run `python src/ -c conf/train/gpt2/retrieval.yml`, I get the following error:

...
File "/Users/ryan/in-context/src/core/function_class.py", line 44, in __init__
    self.p_dist: Distribution = self._init_param_dist()
  File "/Users/ryan/in-context/src/function_classes/retrieval.py", line 12, in _init_param_dist
    return CombinedDistribution(
  File "/Users/ryan/in-context/src/utils.py", line 18, in __init__
    assert dist.batch_shape == batch_shape, \
AssertionError: Batch shapes do not match between all distributions! torch.Size([64, 64]) != torch.Size([]) for Categorical(probs: torch.Size([64]))

Reply:
Ah, this was a missed `.expand` call. Fixed now